# Imports

In [ ]:
# mounting

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from google.colab import files
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [ ]:
# reading data file and peeking in 10 entries

train_data = pd.read_csv('/content/drive/MyDrive/Datasets/Credit_card_fraud/fraudTrain.csv')
train_data.head(10)

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   
5           5   2019-01-01 00:04:08  4767265376804500   
6           6   2019-01-01 00:04:42    30074693890476   
7           7   2019-01-01 00:05:08  6011360759745864   
8           8   2019-01-01 00:05:18  4922710831011201   
9           9   2019-01-01 00:06:01  2720830304681674   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   
5    fraud_Stroman, Hudson and Erdman  gas_transport   94.63   Jennifer   
6               fraud_Rowe-Vandervort    grocery_net   44.54     Kelsey   
7                fraud_Corwin-Collins  gas_transport   71.65     Steven   
8                    fraud_Herzog Ltd       misc_pos    4.27    Heather   
9   fraud_Schoen, Kuphal and Nitzsche    grocery_pos  198.39    Melissa   

       last gender                          street  ...      lat      long  \
0     Banks      F                  561 Perry Cove  ...  36.0788  -81.1781   
1      Gill      F    43039 Riley Greens Suite 393  ...  48.8878 -118.2105   
2   Sanchez      M        594 White Dale Suite 530  ...  42.1808 -112.2620   
3     White      M     9443 Cynthia Court Apt. 038  ...  46.2306 -112.1138   
4    Garcia      M                408 Bradley Rest  ...  38.4207  -79.4629   
5    Conner      F               4655 David Island  ...  40.3750  -75.2045   
6  Richards      F     889 Sarah Station Suite 624  ...  37.9931 -100.9893   
7  Williams      M       231 Flores Pass Suite 720  ...  38.8432  -78.6003   
8     Chase      F     6888 Hicks Stream Suite 954  ...  40.3359  -79.6607   
9   Aguilar      F  21326 Taylor Squares Suite 708  ...  36.5220  -87.3490   

   city_pop                                job         dob  \
0      3495          Psychologist, counselling  1988-03-09   
1       149  Special educational needs teacher  1978-06-21   
2      4154        Nature conservation officer  1962-01-19   
3      1939                    Patent attorney  1967-01-12   
4        99     Dance movement psychotherapist  1986-03-28   
5      2158                  Transport planner  1961-06-19   
6      2691                    Arboriculturist  1993-08-16   
7      6018               Designer, multimedia  1947-08-21   
8      1472          Public affairs consultant  1941-03-07   
9    151785                        Pathologist  1974-03-28   

                          trans_num   unix_time  merch_lat  merch_long  \
0  0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704 -112.154481   
3  6b849c168bdad6f867558c3793159a81  1325376076  47.034331 -112.561071   
4  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999  -78.632459   
5  189a841a0a8ba03058526bcfe566aab5  1325376248  40.653382  -76.152667   
6  83ec1cc84142af6e2acf10c44949e720  1325376282  37.162705 -100.153370   
7  6d294ed2cc447d2c71c7171a3d54967c  1325376308  38.948089  -78.540296   
8  fc28024ce480f8ef21a32d64c93a29f5  1325376318  40.351813  -79.958146   
9  3b9014ea8fb80bd65de0b1463b00b00e  1325376361  37.179198  -87.485381   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  

[10 rows x 23 columns]

In [ ]:
# remane 'unnamed' as 'id'

train_data.rename(columns={'Unnamed: 0':'id'}, inplace= True)

In [ ]:
# getting shape, info and finding missing values from dataset

print("shape of data R:C",train_data.shape)
print(train_data.info())
print('------------------Missing Values--------------------')
missing_values = train_data.isnull().sum()
print(missing_values)

shape of data R:C (1296675, 23)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1296675 entries, 0 to 1296674
Data columns (total 23 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   id                     1296675 non-null  int64  
 1   trans_date_trans_time  1296675 non-null  object 
 2   cc_num                 1296675 non-null  int64  
 3   merchant               1296675 non-null  object 
 4   category               1296675 non-null  object 
 5   amt                    1296675 non-null  float64
 6   first                  1296675 non-null  object 
 7   last                   1296675 non-null  object 
 8   gender                 1296675 non-null  object 
 9   street                 1296675 non-null  object 
 10  city                   1296675 non-null  object 
 11  state                  1296675 non-null  object 
 12  zip                    1296675 non-null  int64  
 13  lat                    1296675 non-null 

In [ ]:
# itrate over all data columns. if datatype of respective column is int or float then fill missing value as mean if datatype is object then fill NA.

# for column in train_data.columns:
#   if train_data[column].dtype == 'int64' or train_data[column].dtype == 'float64':
#     train_data[column].fillna(train_data[column].mean(), inplace=True)
#   elif train_data[column].dtype == 'object':
#     train_data[column].fillna('NA', inplace=True)


In [ ]:
# create 2 separate columns for date and time.

train_data[['trans_date', 'trans_time']] = train_data['trans_date_trans_time'].str.split(' ',expand=True)

#save a copy for any experiments. don't change our original dataset.
copy_data = train_data.copy()

train_data.drop(columns=['trans_date_trans_time'], inplace=True)
train_data.head(5)

id            cc_num                            merchant       category  \
0   0  2703186189652095          fraud_Rippin, Kub and Mann       misc_net   
1   1      630423337322     fraud_Heller, Gutmann and Zieme    grocery_pos   
2   2    38859492057661                fraud_Lind-Buckridge  entertainment   
3   3  3534093764340240  fraud_Kutch, Hermiston and Farrell  gas_transport   
4   4   375534208663984                 fraud_Keeling-Crist       misc_pos   

      amt      first     last gender                        street  \
0    4.97   Jennifer    Banks      F                561 Perry Cove   
1  107.23  Stephanie     Gill      F  43039 Riley Greens Suite 393   
2  220.11     Edward  Sanchez      M      594 White Dale Suite 530   
3   45.00     Jeremy    White      M   9443 Cynthia Court Apt. 038   
4   41.96      Tyler   Garcia      M              408 Bradley Rest   

             city  ... city_pop                                job  \
0  Moravian Falls  ...     3495          Psychologist, counselling   
1          Orient  ...      149  Special educational needs teacher   
2      Malad City  ...     4154        Nature conservation officer   
3         Boulder  ...     1939                    Patent attorney   
4        Doe Hill  ...       99     Dance movement psychotherapist   

          dob                         trans_num   unix_time  merch_lat  \
0  1988-03-09  0b242abb623afc578575680df30655b9  1325376018  36.011293   
1  1978-06-21  1f76529f8574734946361c461b024d99  1325376044  49.159047   
2  1962-01-19  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704   
3  1967-01-12  6b849c168bdad6f867558c3793159a81  1325376076  47.034331   
4  1986-03-28  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999   

   merch_long is_fraud  trans_date  trans_time  
0  -82.048315        0  2019-01-01    00:00:18  
1 -118.186462        0  2019-01-01    00:00:44  
2 -112.154481        0  2019-01-01    00:00:51  
3 -112.561071        0  2019-01-01    00:01:16  
4  -78.632459        0  2019-01-01    00:03:06  

[5 rows x 24 columns]

In [ ]:
# just restructure data file

cols = ['id', 'trans_date', 'trans_time'] + list(train_data.columns)[3:]
train_data = train_data[cols]
train_data.head()

id  trans_date  trans_time       category     amt      first     last  \
0   0  2019-01-01    00:00:18       misc_net    4.97   Jennifer    Banks   
1   1  2019-01-01    00:00:44    grocery_pos  107.23  Stephanie     Gill   
2   2  2019-01-01    00:00:51  entertainment  220.11     Edward  Sanchez   
3   3  2019-01-01    00:01:16  gas_transport   45.00     Jeremy    White   
4   4  2019-01-01    00:03:06       misc_pos   41.96      Tyler   Garcia   

  gender                        street            city  ... city_pop  \
0      F                561 Perry Cove  Moravian Falls  ...     3495   
1      F  43039 Riley Greens Suite 393          Orient  ...      149   
2      M      594 White Dale Suite 530      Malad City  ...     4154   
3      M   9443 Cynthia Court Apt. 038         Boulder  ...     1939   
4      M              408 Bradley Rest        Doe Hill  ...       99   

                                 job         dob  \
0          Psychologist, counselling  1988-03-09   
1  Special educational needs teacher  1978-06-21   
2        Nature conservation officer  1962-01-19   
3                    Patent attorney  1967-01-12   
4     Dance movement psychotherapist  1986-03-28   

                          trans_num   unix_time  merch_lat  merch_long  \
0  0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704 -112.154481   
3  6b849c168bdad6f867558c3793159a81  1325376076  47.034331 -112.561071   
4  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999  -78.632459   

  is_fraud  trans_date  trans_time  
0        0  2019-01-01    00:00:18  
1        0  2019-01-01    00:00:44  
2        0  2019-01-01    00:00:51  
3        0  2019-01-01    00:01:16  
4        0  2019-01-01    00:03:06  

[5 rows x 24 columns]

In [ ]:
# save and download a bit prepared file

# Save the DataFrame to a CSV file on Google Drive.
# train_data.to_csv('/content/drive/MyDrive/Datasets/Credit_card_fraud/train_data_cleaned.csv')

# Download the CSV file from Google Drive.
# files.download('/content/drive/MyDrive/Datasets/Credit_card_fraud/train_data_cleaned.csv')


In [ ]:
# create a saperate list for leagal and fraud trans

legal = train_data[train_data.is_fraud == 0]
fraud = train_data[train_data.is_fraud == 1]

print(legal.shape)
print(fraud.shape)

(1289169, 24)
(7506, 24)


This dataset is highly unbalanced because to have only 7506 from 12891690. if do  further process with this data it might we get 90% as safe trans due to 1289169 safe trans. we need to fist balance this dataset

Under-Sampling

build a sample dataset which contain normal  destribution of legal and fraud

In [ ]:
# taking random sampl form legal data

legal_sample = legal.sample(n=7506)

In [ ]:
# concate leaga_sample and fraud in one df called balanced_df

balenced_df = pd.concat([legal_sample, fraud], axis=0)
balenced_df.head(5)


id  trans_date  trans_time        category     amt      first  \
594742    594742  2019-09-09    07:56:01        misc_net    6.77     Angela   
424023    424023  2019-07-09    07:22:13   gas_transport   47.53  Mackenzie   
156213    156213  2019-03-24    15:52:19  health_fitness    4.62       Jose   
1009359  1009359  2020-02-19    13:54:19            home  128.22     Jordan   
1142442  1142442  2020-04-20    23:10:30   personal_care   48.75      Sarah   

            last gender                        street              city  ...  \
594742      West      F             9471 Wong Islands         Mansfield  ...   
424023     Smith      F             7178 Willis Drive            Hooper  ...   
156213   Vasquez      M           572 Davis Mountains      Lake Jackson  ...   
1009359      May      M             1626 Susan Course             Aledo  ...   
1142442    Clark      F  25961 Beverly Union Apt. 042  North Wilkesboro  ...   

        city_pop                   job         dob  \
594742     92985    Press photographer  1968-05-29   
424023      1745  Engineer, automotive  1973-12-26   
156213     28739        Futures trader  1999-12-27   
1009359    13602           Optometrist  1984-07-05   
1142442    21134   Dispensing optician  1984-03-06   

                                trans_num   unix_time  merch_lat merch_long  \
594742   b9998b5bff2549b365b045ff8f532c2c  1347177361  41.002086 -82.337985   
424023   8aeb7e3a83a1c8fb1fbee1237f71343b  1341818533  40.854694 -96.270041   
156213   ed87a95825edd18fcae3e3d0a8b7feda  1332604339  29.967787 -95.636365   
1009359  3a9976ff07ff45ccdcfdf3cb339ab382  1361282059  32.721423 -96.800355   
1142442  f8b2f49fe5bc925fd4e7244737cb2629  1366499430  37.050564 -80.922832   

        is_fraud  trans_date  trans_time  
594742         0  2019-09-09    07:56:01  
424023         0  2019-07-09    07:22:13  
156213         0  2019-03-24    15:52:19  
1009359        0  2020-02-19    13:54:19  
1142442        0  2020-04-20    23:10:30  

[5 rows x 24 columns]

In [ ]:
# new balanced data sheet is ready and now time to get value count.
balenced_df['is_fraud'].value_counts()

is_fraud
0    7506
1    7506
Name: count, dtype: int64

In [ ]:
balenced_df.shape

(15012, 24)

Here we got balenced dataframe

amt, city_pop, dob , trans_num,

In [ ]:
# # just cheking test file.

# tst_file =pd.read_csv('/content/drive/MyDrive/Datasets/Credit_card_fraud/fraudTrain.csv')
# tst_file.head(5)

# tst_file.shape

Splitting data feature and targate


In [ ]:
X = balenced_df.drop(columns=['is_fraud'], axis=1)
Y = balenced_df['is_fraud']

#X1 gives batter accuracy score
X1= balenced_df[['amt','city_pop',]]
#
#X2 increases accuracy by 1
X2= balenced_df[['id','amt','city_pop']]

In [ ]:
print(X2)

              id      amt  city_pop
594742    594742     6.77     92985
424023    424023    47.53      1745
156213    156213     4.62     28739
1009359  1009359   128.22     13602
1142442  1142442    48.75     21134
...          ...      ...       ...
1295399  1295399   977.01    105638
1295491  1295491  1210.91    105638
1295532  1295532    10.24     71335
1295666  1295666    21.69        23
1295733  1295733    10.20     71335

[15012 rows x 3 columns]


In [ ]:
print(Y)

594742     0
424023     0
156213     0
1009359    0
1142442    0
          ..
1295399    1
1295491    1
1295532    1
1295666    1
1295733    1
Name: is_fraud, Length: 15012, dtype: int64


Split data into training and test

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X2, Y, test_size=0.2, stratify=Y, random_state=2)

In [ ]:
print(X2.shape, X_train.shape, X_test.shape)

(15012, 3) (12009, 3) (3003, 3)


Model Training

In [ ]:
lr_model = LogisticRegression()
lr_model.fit(X_train , y_train)

LogisticRegression()

Evalution

Acuuracy Score

In [ ]:
X_train_pred = lr_model.predict(X_train)
X_pred_acc = accuracy_score(X_train_pred, y_train)

In [ ]:
print("Accuracy Score: " , X_pred_acc)

Accuracy Score:  0.7917395286868182


In [ ]:
# test prediction

X_test_pred = lr_model.predict(X_test)
X_test_acc = accuracy_score(X_test_pred, y_test)

In [ ]:
print("Test Accuracy Score: ", X_test_acc)

Test Accuracy Score:  0.7848817848817848
